### PR TITLE
Add get_redirect_to_lms_login_url helper

### PR DESCRIPTION
**Pull request by shadinaif**
## Change description

Add `get_redirect_to_lms_login_url` to be used by our [theme](https://github.com/appsembler/edx-theme-codebase/pull/225) to redirect to LMS login if needed

**Example scenario before this change:**
* Studio is not logged in yet
* An author clicks (View Unit in Studio) in LMS <LMS logged in for sure>
* Redirects to Studio Login page. The user must follow instructions and open studio though LMS

**Same scenario after this change:**
* Studio is not logged in yet
* An author clicks (View Unit in Studio) in LMS <LMS logged in for sure>
* Redirects to Studio Login page
* Studio fetches the `course_id` from `next` argument in the URL
* Now studio knows the related site: redirects to `https://the_site/studio/?next=***`
* Auto magic-link performed, then a redirect to the originally requested page is performed

## Type of change
- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

## Related issues

Related to: https://appsembler.atlassian.net/browse/RED-3242
Required by: https://github.com/appsembler/edx-theme-codebase/pull/225

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: openedx/core/djangoapps/appsembler/tahoe_idp/helpers.py [hunk 1]
```diff
@@ -185,3 +193,61 @@ def is_studio_login_form_overridden():
...
+    if re.match(r'^/container/block-v1:.+\+type@.+$', url):
+        return re.sub(r'^/container/block-v1:(.+)\+type@.+$', r'course-v1:\1', url)
+
+    # Matching for url containing course-v1 course_id in the middle
+    elif re.match(r'^.+/course-v1:.+\+.+/.*$', url):
+        return re.sub(r'^.+/course-v1:(.+)/.*', r'course-v1:\1', url)
+
+    # Matching for url containing course-v1 course_id at the end
+    elif re.match(r'^.+/course-v1:.+\+.+$', url):
+        return re.sub(r'^.+/course-v1:(.+)', r'course-v1:\1', url)
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/tahoe_idp/helpers.py:217]: > Some people, when confronted with a problem, think "I know, I'll use regular expressions." Now they have two problems. -- [CodingHorror](https://blog.codinghorror.com/regular-expressions-now-you-have-two-problems/)


Snarkiness aside `course-v1:(.+)` is the wrong pattern to match against. It needs to be a more explicit pattern: `course-v1:([^+]+)` so it doesn't eagerly catch other plusses. Or better, it should be the exact edX patterns for course keys:

 - https://github.com/openedx/opaque-keys/blob/master/opaque_keys/edx/locator.py

File: openedx/core/djangoapps/appsembler/tahoe_idp/tests/test_tahoe_idp_helpers.py [hunk 2]
```diff
@@ -232,3 +244,83 @@ def test_is_studio_login_form_overridden_flag_available(flag_value, expected_res
     """
     with patch.dict('django.conf.settings.FEATURES', {'TAHOE_IDP_STUDIO_LOGIN_FORM_OVERRIDE': flag_value}):
         assert helpers.is_studio_login_form_overridden() is expected_result
+
+
+@pytest.mark.parametrize('url,expected_result', [
+    ('https://example.com/some_path/course-v1:ORG+org_2022', 'course-v1:ORG+org_2022'),
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/tahoe_idp/tests/test_tahoe_idp_helpers.py:250]: Course IDs are always `course-v1:org+course+run` e.g. `course-v1:redis+cache+2022_Sprint`. I think both of the code and test cases are incorrect.

File: openedx/core/djangoapps/appsembler/tahoe_idp/helpers.py [hunk 3]
```diff
@@ -185,3 +193,61 @@ def is_studio_login_form_overridden():
     if settings.FEATURES.get('TAHOE_IDP_STUDIO_LOGIN_FORM_OVERRIDE', None):
         return True
     return False
+
+
+def extract_course_id_from_url(url):
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/tahoe_idp/helpers.py:198]: That would be an easier task to do.

```suggestion
def extract_org_short_name_from_url(url):
```

File: openedx/core/djangoapps/appsembler/tahoe_idp/helpers.py [hunk 4]
```diff
@@ -185,3 +193,61 @@ def is_studio_login_form_overridden():
...
+    :return: redirect url if a valid course key found, otherwise return empty string
+    """
+    if not (request and request.GET.get('next')):
+        return ''
+
+    next_url = request.GET['next']
+    course_id = extract_course_id_from_url(next_url)
+
+    try:
+        organization = get_organization_by_course(course_id)
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/tahoe_idp/helpers.py:239]:
```suggestion
        organization = Organization.objects.get(short_name=org_short_name)
```

**Review by OmarIthawi — CHANGES_REQUESTED** [review]
Thanks Shadi!!

**Required change:** I think the code is designed to work with the _incorrect_ course IDs.

Course IDs are always `course-v1:org+course+run` e.g. `course-v1:redis+cache+2022_Sprint`. I think both of the code and test cases are incorrect.

**Suggested change:** Try to get the organization ID instead of the course ID. If you think this is a good idea.

File: openedx/core/djangoapps/appsembler/tahoe_idp/helpers.py (hunk 1 above)
Comment (shadinaif) [reply openedx/core/djangoapps/appsembler/tahoe_idp/helpers.py]: Done, thank you!

**Comment by shadinaif** [comment]
I've refactored the code @OmarIthawi , please take a look now 🌷

**Review by OmarIthawi — APPROVED** [review]
Thanks @shadinaif for going through this RegExp mess! I think it's much better now.

Tests also looks really good!

**Pull request merged**